lib/xnvme_be_macos_driverkit_dev.c

```diff
@@ -0,0 +1,231 @@
+	}
+
+	qid--;
+
+	XNVME_DEBUG("INFO: nvme_create_ioqpair(%d, %d, %d)", qid, qd + 1, flags);
+
+	// nvme queue capacity must be one larger than the requested capacity
+	// since only n-1 slots in an NVMe queue may be used
+	ioqpair_create = (NvmeQueue){
+		.id = (uint64_t)qid,
+		.depth = (uint64_t)qd + 1,
+		.vector = -1,
+		.flags = (uint64_t)flags,
+	};
+	output_cnt = sizeof(NvmeQueue);
+
+	ret = IOConnectCallStructMethod(state->device_connection, NVME_CREATE_QUEUE_PAIR,
+					&ioqpair_create, sizeof(NvmeQueue), &ioqpair_return,
+					&output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallStructMethod(NVME_CREATE_QUEUE_PAIR); "
+			    "ret(0x%08x), '%s'",
+			    ret, mach_error_string(ret));
+		return -EAGAIN;
+	}
+	XNVME_DEBUG("INFO: IOConnectCallStructMethod(NVME_CREATE_QUEUE_PAIR); success");
+
+	state->qid_sync = ioqpair_return.id;
+	state->qidmap &= ~(1ULL << qid);
+
+	return qid;
+}
+
+int
+_xnvme_be_driverkit_delete_ioqpair(struct xnvme_be_macos_driverkit_state *state, int qid)
+{
+	size_t output_cnt = 0;
+	NvmeQueue ioqpair_delete = {
+		.id = (uint64_t)qid,
+	};
+	kern_return_t ret;
+
+	ret = IOConnectCallStructMethod(state->device_connection, NVME_DELETE_QUEUE_PAIR,
+					&ioqpair_delete, sizeof(NvmeQueue), NULL, &output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallStructMethod(NVME_DELETE_QUEUE_PAIR); "
+			    "ret(0x%08x), '%s'",
+			    ret, mach_error_string(ret));
+		return -EIO;
+	}
+	XNVME_DEBUG("INFO: IOConnectCallStructMethod(NVME_DELETE_QUEUE_PAIR); success;"
+		    "ret(0x%08x), '%s'",
+		    ret, mach_error_string(ret));
+
+	state->qidmap |= 1ULL << qid;
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_dev_open(struct xnvme_dev *dev)
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+	int qpid;
+	kern_return_t err;
+	io_service_t service;
+	CFMutableDictionaryRef matching_dict;
+	io_connect_t connection;
+
+	matching_dict = IOServiceNameMatching(dev->ident.uri);
+	service = IOServiceGetMatchingService(kIOMasterPortDefault, matching_dict);
+	if (!service) {
+		XNVME_DEBUG("FAILED: IOServiceGetMatchingService()");
+		return -ENODEV;
+	}
+
+	err = IOServiceOpen(service, mach_task_self_, MACVFN_CLIENT_VERSION, &connection);
+	if (err != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOServiceOpen(MACVFN_CLIENT_VERSION); connection(%u),"
+			    "ret(0x%08x), '%s'",
+			    connection, err, mach_error_string(err));
+		return -EIO;
+	}
+	XNVME_DEBUG("INFO: IOServiceOpen(MACVFN_CLIENT_VERSION); connection: %u", connection);
+
+	state->device_service = service;
+	state->device_connection = connection;
+
+	err = IOConnectCallScalarMethod(state->device_connection, NVME_INIT, NULL, 0, NULL, 0);
+	if (err != kIOReturnSuccess) {
+		XNVME_DEBUG("FAILED: IOConnectCallScalarMethod(NVME_INIT);"
+			    "ret(0x%08x), '%s'",
+			    err, mach_error_string(err));
+		return -EIO;
+	}
+	XNVME_DEBUG("NVME_INIT done\n");
```

Comment:
Should be: ``XNVME_DEBUG("INFO: NVME_INIT done\n");``

Reply:
Done, but without the newline